File: bindings/wasm/examples/src/0_basic/3_deactivate_did.ts
// Copyright 2020-2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { IotaDocument, IotaIdentityClient, JwkMemStore, KeyIdMemStore, Storage } from "@iota/identity-wasm/node";
import { AliasOutput, Client, IRent, MnemonicSecretManager, Utils } from "@iota/sdk-wasm/node";
import { API_ENDPOINT, createDid } from "../util";

/** Demonstrates how to deactivate a DID in an Alias Output. */
export async function deactivateIdentity() {
    const client = new Client({
        primaryNode: API_ENDPOINT,
        localPow: true,
    });
    const didClient = new IotaIdentityClient(client);

    // Generate a random mnemonic for our wallet.
    const secretManager: MnemonicSecretManager = {
        mnemonic: Utils.generateMnemonic(),
    };

    // Creates a new wallet and identity (see "0_create_did" example).
    const storage: Storage = new Storage(new JwkMemStore(), new KeyIdMemStore());
    let { document } = await createDid(
        client,
        secretManager,
        storage,
    );
    const did = document.id();

    // Resolve the latest state of the DID document, so we can reactivate it later.
    // Technically this is equivalent to the document above.
    document = await didClient.resolveDid(did);

    // Deactivate the DID by publishing an empty document.
    // This process can be reversed since the Alias Output is not destroyed.
    // Deactivation may only be performed by the state controller of the Alias Output.
    let deactivatedOutput: AliasOutput = await didClient.deactivateDidOutput(did);

    // Optional: reduce and reclaim the storage deposit, sending the tokens to the state controller.
    const rentStructure: IRent = await didClient.getRentStructure();

    deactivatedOutput = await client.buildAliasOutput({
        ...deactivatedOutput,
        amount: Utils.computeStorageDeposit(deactivatedOutput, rentStructure),
        aliasId: deactivatedOutput.getAliasId(),
        unlockConditions: deactivatedOutput.getUnlockConditions(),
    });

    // Publish the deactivated DID document.
    await didClient.publishDidOutput(secretManager, deactivatedOutput);

    // Resolving a deactivated DID returns an empty DID document
    // with its `deactivated` metadata field set to `true`.
    let deactivated: IotaDocument = await didClient.resolveDid(did);
    console.log("Deactivated DID document:", JSON.stringify(deactivated, null, 2));
    if (deactivated.metadataDeactivated() !== true) {
        throw new Error("Failed to deactivate DID document");
    }

    // Re-activate the DID by publishing a valid DID document.
    let reactivatedOutput: AliasOutput = await didClient.updateDidOutput(document);

    // Increase the storage deposit to the minimum again, if it was reclaimed during deactivation.
    reactivatedOutput = await client.buildAliasOutput({
        ...reactivatedOutput,
        amount: Utils.computeStorageDeposit(reactivatedOutput, rentStructure),
        aliasId: reactivatedOutput.getAliasId(),
        unlockConditions: reactivatedOutput.getUnlockConditions(),
    });

    await didClient.publishDidOutput(secretManager, reactivatedOutput);

    // Resolve the reactivated DID document.
    let reactivated: IotaDocument = await didClient.resolveDid(did);
    if (reactivated.metadataDeactivated() === true) {
        throw new Error("Failed to reactivate DID document");
    }
}

File: examples/kinesis/0_basic/3_deactivate_did.rs
// Copyright 2020-2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use examples_kinesis::create_kinesis_did_document;
use examples_kinesis::get_client_and_create_account;
use examples_kinesis::get_memstorage;
use examples_kinesis::TEST_GAS_BUDGET;
use identity_iota::iota::IotaDID;
use identity_iota::iota::IotaDocument;

/// Demonstrates how to deactivate a DID in an Alias Output.
#[tokio::main]
async fn main() -> anyhow::Result<()> {
  // create new client to interact with chain and get funded account with keys
  let storage = get_memstorage()?;
  let identity_client = get_client_and_create_account(&storage).await?;
  // create new DID document and publish it
  let (document, _) = create_kinesis_did_document(&identity_client, &storage).await?;
  let did: IotaDID = document.id().clone();

  // Deactivate the DID by publishing an empty document.
  // This process can be reversed since the Alias Output is not destroyed.
  // Deactivation may only be performed by the state controller of the Alias Output.
  identity_client.deactivate_did_output(&did, TEST_GAS_BUDGET).await?;

  // Resolving a deactivated DID returns an empty DID document
  // with its `deactivated` metadata field set to `true`.
  let deactivated: IotaDocument = identity_client.resolve_did(&did).await?;
  println!("Deactivated DID document: {deactivated:#}");
  assert_eq!(deactivated.metadata.deactivated, Some(true));

  // Re-activate the DID by publishing a valid DID document.
  let reactivated: IotaDocument = identity_client
    .publish_did_document_update(document.clone(), TEST_GAS_BUDGET)
    .await?;
  println!("Reactivated DID document result: {reactivated:#}");

  let resolved: IotaDocument = identity_client.resolve_did(&did).await?;
  println!("Reactivated DID document resolved from chain: {resolved:#}");

  Ok(())
}
